cpp/include/raft/core/operators.hpp

```diff
@@ -0,0 +1,292 @@
+  {
+    return a + b;
+  }
+};
+
+struct sub_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(const T1& a, const T2& b) const
+  {
+    return a - b;
+  }
+};
+
+struct mul_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(const T1& a, const T2& b) const
+  {
+    return a * b;
+  }
+};
+
+struct div_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(const T1& a, const T2& b) const
+  {
+    return a / b;
+  }
+};
+
+struct div_checkzero_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    if (b == Type{0}) { return Type{0}; }
+    return a / b;
+  }
+};
+
+struct pow_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    return std::pow(a, b);
+  }
+};
+
+struct min_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    if (a > b) { return b; }
+    return a;
+  }
+};
+
+struct max_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    if (b > a) { return b; }
+    return a;
+  }
+};
+
+struct sqdiff_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(const Type& a, const Type& b) const
+  {
+    Type diff = a - b;
+    return diff * diff;
+  }
+};
+
+struct argmin_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+struct argmax_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+template <typename ScalarT>
+struct const_op {
+  const ScalarT scalar;
+
+  constexpr const_op(const ScalarT& s) : scalar{s} {}
+
+  template <typename... Args>
+  constexpr RAFT_INLINE_FUNCTION ScalarT operator()(Args...) const
+  {
+    return scalar;
+  }
+};
+
+template <typename ScalarT, typename BinaryOpT>
+struct scalar_op {
+  const ScalarT scalar;
+  const BinaryOpT composed_op;
+
+  template <typename OpT     = BinaryOpT,
+            typename UnusedT = std::enable_if_t<std::is_default_constructible_v<OpT>>>
+  constexpr scalar_op(const ScalarT& s) : scalar{s}, composed_op{}
+  {
+  }
+  constexpr scalar_op(const ScalarT& s, BinaryOpT o) : scalar{s}, composed_op{o} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(InT a) const
+  {
+    return composed_op(a, scalar);
+  }
+};
+
+template <typename Type>
+using scalar_add_op = scalar_op<Type, add_op>;
```

Comment:
What would you think about renaming these `scalar_xxx_op` into something like `xxx_const_op`? I see the intuition of using the word scalar since we're using these operators mostly in kernels that map vectors and matrices. Yet, if we look at the `operators.hpp` as an independent piece of the api, the word `const` seems to fit better, since all of the operators here are, in fact, defined on scalars :) .
I'm not insisting on this, though.

Reply:
Then, perhaps, it would also make sense to rename and slightly generalize `scalar_op` to something like this:
```C++
template <size_t ArgPos, typename ConstT, typename MultiArgOpT>
struct plug_const_op {...}
```
(or just hide it into `detail` namespace)

Reply:
When considering the operation at the level of the primitive you're using it for, you have a vector or matrix on the left hand, and a scalar on the right hand. I think the term scalar is well understood in that context.

Of course, it's a matter of personal preference and mine is to keep the `scalar` names, since there is also a `const_op` that returns a constant no matter the input. But I don't have a strong opinion on this, so if others weigh in on this I might change my mind :)

Reply:
> When considering the operation at the level of the primitive you're using it for, you have a vector or matrix on the left hand, and a scalar on the right hand. I think the term scalar is well understood in that context.

Sure, but my point here is that, since this is going to be a part of public api, the users are not obliged to pass these operators to `unary_op` , `matrix_vector_op`, etc. They could pass it somewhere less obvious, or technically even use them on the host side directly.

Reply:
I don't like the idea that these ops could be used directly, I think they should only be used as arguments to functor-based prims.

```c++
c = raft::add{}(a, b);  // don't do this!
c = a + b;              // do this
```

Reply:
Alright, I am outnumbered. I've renamed accordingly.